tests/honey_badger.rs

```diff
@@ -14,149 +13,165 @@ use rand::{seq::SliceRandom, Rng};
-    pub fn new(
-        num_good: usize,
-        num_adv: usize,
-        adv_nodes: BTreeMap<NodeId, Arc<NetworkInfo<NodeId>>>,
-        scheduler: MessageScheduler,
-    ) -> FaultyShareAdversary {
+    /// Creates a new adversary with the necessary network info instances
+    pub fn new(netinfo_mutex: Arc<Mutex<NetworkInfoMap>>) -> Self {
         FaultyShareAdversary {
-            num_good,
-            num_adv,
-            scheduler,
             share_triggers: BTreeMap::new(),
-            adv_nodes,
+            netinfo_mutex,
         }
     }
 }
 
 impl Adversary<UsizeHoneyBadger> for FaultyShareAdversary {
-    fn pick_node(&self, nodes: &BTreeMap<NodeId, TestNode<UsizeHoneyBadger>>) -> NodeId {
-        self.scheduler.pick_node(nodes)
+    #[inline]
+    fn pre_crank<R: Rng>(
+        &mut self,
+        mut net: NetMutHandle<'_, UsizeHoneyBadger, Self>,
+        rng: &mut R,
+    ) {
+        sort_by_random_node(&mut net, rng);
     }
 
-    fn push_message(
+    #[inline]
+    fn tamper<R: Rng>(
         &mut self,
-        sender_id: NodeId,
-        msg: TargetedMessage<<UsizeHoneyBadger as ConsensusProtocol>::Message, NodeId>,
-    ) {
-        let NodeId(sender_id) = sender_id;
-        if sender_id < self.num_good {
-            if let TargetedMessage {
-                target: Target::All,
-                message: sender_queue::Message::Algo(hb_msg),
-            } = msg
-            {
-                let epoch = hb_msg.epoch();
-                // Set the trigger to simulate decryption share messages.
-                self.share_triggers.entry(epoch).or_insert(true);
-            }
+        mut net: NetMutHandle<'_, UsizeHoneyBadger, Self>,
+        msg: HoneyBadgerMessage,
+        rng: &mut R,
+    ) -> Result<CpStep<UsizeHoneyBadger>, CrankError<UsizeHoneyBadger>> {
+        if let sender_queue::Message::Algo(hb_msg) = msg.payload() {
+            let epoch = hb_msg.epoch();
+            // Set the trigger to simulate decryption share messages
+            // if epoch has not been encountered yet.
+            self.share_triggers.entry(epoch).or_insert(true);
         }
-    }
 
-    fn step(&mut self) -> Vec<MessageWithSender<UsizeHoneyBadger>> {
-        let mut outgoing = vec![];
-        let fake_proposal = &Vec::from("X marks the spot");
+        let mut step = net.dispatch_message(msg, rng)?;
 
+        let fake_proposal = &Vec::from("X marks the spot");
+        // For each untriggered epoch, send fake shares
         for (epoch, trigger_set) in &mut self.share_triggers {
             if *trigger_set {
                 // Unset the trigger.
                 *trigger_set = false;
+
+                // Get node id vectors up-front to avoid borrow issues
+                let faulty_node_ids: Vec<NodeId> =
+                    net.faulty_nodes_mut().map(|node| *node.id()).collect();
+                let all_node_ids: Vec<NodeId> = net.nodes_mut().map(|node| node.id()).collect();
```

Comment:
Is that closure needed?
```rust
.map(Node::id)
```

Reply:
It does expect a function taking a &NodeMutHandle as first argument, this gets close:
`.map(NodeMutHandle::id)`

but not quite, it has an extra lifetime parameter, so it boils down to the required:
`fn(net::adversary::NodeMutHandle<'_, hbbft::sender_queue::SenderQueue<hbbft::honey_badger::honey_badger::HoneyBadger<std::vec::Vec<usize>, u16>>>) -> _
`
vs NodeMutHandle's:
`for<'r> fn(&'r net::adversary::NodeMutHandle<'_, hbbft::sender_queue::SenderQueue<hbbft::honey_badger::honey_badger::HoneyBadger<std::vec::Vec<usize>, u16>>>) -> _`